### DISCUSSION TITLE
Spiking outside temp

### DISCUSSION THREAD

**Discussion by razorclam** [gekkekoe/esphome-ecodan-hp#92] [Announcements]
Hello again!
I have noticed regular spiking in the outside temp. Could this be something caused by the ESP or more likey a faulty sensor or someting else?

![image](https://github.com/user-attachments/assets/a562a7b0-2be2-47a4-be11-b2a9d8308730)

**Comment by merlin051** [comment]
Defrost cycles causing evaporation that which makes the ambient temperature
spike temporarily

**Reply by razorclam** [reply]
Thanks for the prompt reply. Never thought of that and makes perfect sense.
However.... checking the operation mode the defrost is happening outside of those times?

![image](https://github.com/user-attachments/assets/6dd28373-dc2f-4b09-bbad-e3cf8e4fa1b9)

**Reply by razorclam** [reply]
Also looks like those spikes are causing the mean flow temp to come down

**Reply by DamijanD** [reply]
Check the defrost sensor (not the mode)

**Reply by razorclam** [reply]
there it is!
![image](https://github.com/user-attachments/assets/3e3201da-5b05-45fa-9927-f01606dc264d)

**Comment by DamijanD** [comment]
I have fork where I apply filtering to this and it works. 
You can check this rejected PR:
https://github.com/gekkekoe/esphome-ecodan-hp/pull/85
If you need this you can use (https://github.com/DamijanD/esphome-ecodan-hp/tree/outsidetemp2)

**Comment by gekkekoe** [comment]
Thanks @DamijanD & @merlin051  for replying, yes it's perfect normal. You can also use it as indicator for a defrost. It is what the sensor observes and it's put there by design, don't worry about it too much. 

I still think it's way easier to filter it out in HA tbh or get a reading from a meteo station near you.

**Reply by DamijanD** [reply]
I probably agree that it is easier to do it in HA (but to be honest I don't know how).  The only "problem" that I see is that in HA everyone needs to do it and if it is filtered in the esp then nobody needs to do it.

**Reply by gekkekoe** [reply]
I prefer the spikes, but when I have some spare time I can have a look at HA. Probably an additional template sensor or filter, looks doable (template sensor that derives from the reguar outside temp and apply logic. or use filter property)

see https://esphome.io/components/sensor/index.html#sensor-filters
https://esphome.io/components/sensor/template.html

template sensor is probably easiest

**Reply by gekkekoe** [reply]
HA seems to have support for outliers:
https://www.home-assistant.io/integrations/filter/#outlier
So create a sensor that derives from the outside temp and define the outlier filter, tweak window and radius

**Comment by gekkekoe** [comment]
@RazorClam why is your unit using 3kW between 1:30 - 6:00. That's looks way too high.
Do you have booster running at those times?

What unit do you have and what compressor frequency was it doing? Also show booster en immersion sensor
You are probably heating during the night because of the cheaper electricity rates?

**Reply by razorclam** [reply]
That my friend is a can of worms of a question....
Our outdoor unit is PUZ-WM85VAA
Our developer in their wisdom installed 22mm chipboard, foam underlay and deep pile carpet upstairs. (And i mean like 18 degrees, thats about the max in rooms with no solar gain)
When it gets down to <7 degrees we need to run constantly at around 45/50 degrees to keep upstairs warm.
Shame because downstairs is screed and LVT and works perfectly.

**Reply by gekkekoe** [reply]
can you show booster sensor, compressor frequency and flow rate for that time interval? Can't you install extra radiators upstairs?

But also why heating at night its not efficient, unless the tarifs are way lower

**Reply by razorclam** [reply]
Our heating is on 24/7 only way to keep upstairs warm. Upstairs thermostats all set to 30. You can see the temps only just about hang on

As requested....
![image](https://github.com/user-attachments/assets/fa4e144c-967a-4ac8-add2-44417bde0e28)
![image](https://github.com/user-attachments/assets/6b722b7d-d488-4d09-8a45-0ba0987fd50b)
![image](https://github.com/user-attachments/assets/28e19a42-9069-4104-901a-e632ed4593f6)

Feedback welcome for sure. Our entire estate is complaining!

**Reply by gekkekoe** [reply]
few observations
- flow rate seems decent
- between 2:00-7:00 compressor frequency maxed out
- no booster element running
- Feed temp and return temp drops between 2:00-7:00

Did you increase the feed temp between 02:00-07:00 ? I think I see that in you compensation curve ? +2c is really a lot for the heatpump. Its better to keep it at 40 the whole day. You can see that you actually are getting the oposite, it drops down to around 35c (this is also really strange!)

I also see ground floor temp increasing after 02:00, but that makes no sense, your feed temp is lower than before.
Do you have automations that turns of certain heating circuits/groups ? It's the only explaination I can come up with. So when you open more circuits and increrase the feed temp, your heatpump cannot deliver that amount at that time. It's best to keep all circuits open and have a stable feed temp. Your heatpump will be most happy that way (and your engery bill). You need to figure out what feed temp that should be.

So have all circquits open. Bring up the feed temp slowly and see where it's comfortable and keep it there. Don't adjust on the fly, +2c is a LOT.

Keep in mind that output power is determined by:
kW = flow rate/60 * (feed temp - return temp) * 4.18 (The estimated power output sensor calculates this)

your unit has a certain amount of max power output given the outside temp (see databook), and the flow rate did not change, so every degree increase coressond to 
30/60 * degrees delta * 4.18 = 2kW (!). 

+2c basically means asking 4kW power for this flow rate, and I think your unit is only doing max 10 kW max at 25c feed temp. At 40c it will be much lower.

**Reply by razorclam** [reply]
Excellent thanks...
I would love to have all circuits open but down stairs would exceed 26 degrees very quickly
No automations just weather comp curve at ecodan, upstairs stats set to 30 and downstairs set to 21.
I was under the impression that the compensation curve is most efficient but your saying at these high temps maybe not?

**Reply by razorclam** [reply]
So far today...
https://grafana.razortelecom.co.uk/public-dashboards/c6f9710e5cea42bd974d0e4f90230d5c?orgId=1&refresh=1m

**Reply by gekkekoe** [reply]
key is to have everything open, AND then you can lower the feed temp (Heatpump will be much happier).
Lower it so that its comfortable downstairs and upstairs and keep it at that feed temp once you have found it.

compensation curve is efficient if set correctly. The curve is nothing more than the feed temp given outside temp where you can maintain a certain temp. Once you figured it out for one outside temp, you can tweak it a bit to find it for other temps and then setup a curve. 

Opening (a lote) more circuits is like requesting a lot of additional output power from the heatpump. You are treating it too much like a traditional high temp heating. It's not, its LOW temperature heating.

It's best to keep the delta T (feed - return) as low as possible and the flow rate as high as possible while having comfortable indoor temps. That is the most efficient way. Those are the only 2 parameters you can tweek (via feed temp, and pump speed, set it to max (5)).

**Reply by razorclam** [reply]
Thanks again for the details.
My issue is that I can run the downstairs at <35 no problem. At the same temp upstairs there is never any warmth, the temps just drop to uncomfortable levels. (+/- 13 degrees)
There is such an imbalance of output from the ground and first floor that I think normal UFH theory goes out the window.
I completely agree about running it as low as possible. I tried this the same week I moved in a year ago. Unfortunately we can only cap the thermostats downstairs and keep raising temp until there is warmth upstairs.
The developer has recently replied to complaints saying we should run the system at fixed flow of 50 degrees between December and March.....
its a nightmare!

**Reply by gekkekoe** [reply]
yeah you need some radiators above, but you can use my guidelines to reduce the energy. Also running at max also triggers faster defrost, so its just a downwards spiral.

**Reply by razorclam** [reply]
You suggest setting pump speed to 5? Any bad effects from that?

**Reply by bwduncan** [reply]
> You suggest setting pump speed to 5? Any bad effects from that?

noise :laughing: I tried changing mine from 3 to 5, flow rate increased from 16 to 20 L/min. Lower dT which reduces output, which is not what I wanted when tariffs are very low for a short time.

I wish there was a way to control pump speed from HA, then I could have high COP when price is high, and high output when price is low.

**Reply by gekkekoe** [reply]
yeah noise, hmm radiators ? or air in the circuit?

If you increase pump only it should increase output if the rest is the same...

**Reply by bwduncan** [reply]
Just the sound of the pump itself :slightly_smiling_face: maybe i'm sensitive, but i can hear the difference in the water pump all through my house.

Maybe I didn't look closely enough, but if flow rate increases by 20% (16 -> 20) and dT reduces by 50% (6 -> 3), that's a decrease in output (6.7 kW -> 4.1 kW), surely? I see what you mean, intuitively it should increase output, but only if cold water is returning to the HP...

**Reply by gekkekoe** [reply]
> Just the sound of the pump itself 🙂 maybe i'm sensitive, but i can hear the difference in the water pump all through my house.
> 
> Maybe I didn't look closely enough, but if flow rate increases by 20% (16 -> 20) and dT reduces by 50% (6 -> 3), that's a decrease in output (6.7 kW -> 4.1 kW), surely? I see what you mean, intuitively it should increase output, but only if cold water is returning to the HP...

I case of rediators, only thing I could think of is that there's a max flow where it can exchange its heat. If we send water too fast, the return temp increases (cause flow was to high to exchange all the heat), and the output can then be decreased. So you need to find sweetspot for your radiators (now I'm assuming you have radiators)

**Reply by razorclam** [reply]
UFH on both ground and first floor. 
at around 8 degrees outside it’s ok at around 35-38 flow. As the outside temp drops to <4 it seems to just fall apart.

**Reply by gekkekoe** [reply]
> UFH on both ground and first floor. at around 8 degrees outside it’s ok at around 35-38 flow. As the outside temp drops to <4 it seems to just fall apart.

Ah missed that part, but then it should be easy?

Maintain flow of 35-38 liters/m.
Tweak feed to so that its comfortable in the whole house.
Leave all circuits open
you might have to tweak the flow the the groups (at the underfloor heating devider) upstairs such that the temp is comfortable.
If it gets too warm upstairs, restrict the flow a bit, test for couple of hours, repeat until the whole house is comfortable.

Your issue is that when its < 4c, it cools down and then you open circquits, asking for a lot of output power. You need to keep the house at constant temp, then at < 4c you will see that it needs way less output to maintain the temp.

when done correctly, you should get something likes this (my flow is around 21-22 L/m, this is the max for my unit):
![image](https://github.com/user-attachments/assets/31425ca3-0e85-432f-9cc2-c9f87f0592f1)

The spike is the DHW that kicked in. For heating it remains low/stable.

**Reply by razorclam** [reply]
All upstairs thermostats are on 30. I think the underlay and carpet is stopping the heat transfer and lower ambient temps in the room. The weather curve increases the flow temp when the outside temp drops but is not transferred into the room. 
I respect what you’re saying about consistent flow temp. I set is at 38 fixed flow for a week and it used quite a bit less energy. Maybe that’s the answer.

**Reply by gekkekoe** [reply]
> All upstairs thermostats are on 30. I think the underlay and carpet is stopping the heat transfer and lower ambient temps in the room. The weather curve increases the flow temp when the outside temp drops but is not transferred into the room. I respect what you’re saying about consistent flow temp. I set is at 38 fixed flow for a week and it used quite a bit less energy. Maybe that’s the answer.

yeah the 22mm is killing the heat exchange upstairs... But there should be some flow / feed temp setting where it will be in balance.

**Reply by bwduncan** [reply]
> I case of rediators, only thing I could think of is that there's a max flow where it can exchange its heat. If we send water too fast, the return temp increases (cause flow was to high to exchange all the heat), and the output can then be decreased. So you need to find sweetspot for your radiators (now I'm assuming you have radiators)

(I do have radiators) Yes that seems likely, I can't remember what the absolute flow temp was when I measured this. It was probably high.

I have long thought that my pump is oversized. It's 8 kW and I have mostly ~1 m K2 rads, one K3 in the kitchen :thinking: . My house is ~80 m². The compressor often shuts off briefly and then cycles on again. When the flow temp is set <35 °C it will do this all day. It can run continuously at >40 °C flow temp but I only do this when electricity is really cheap :laughing: 

Anyway I think I am diluting this thread so I will butt out :slightly_smiling_face:

**Reply by gekkekoe** [reply]
> > I case of rediators, only thing I could think of is that there's a max flow where it can exchange its heat. If we send water too fast, the return temp increases (cause flow was to high to exchange all the heat), and the output can then be decreased. So you need to find sweetspot for your radiators (now I'm assuming you have radiators)
> 
> (I do have radiators) Yes that seems likely, I can't remember what the absolute flow temp was when I measured this. It was probably high.
> 
> I have long thought that my pump is oversized. It's 8 kW and I have mostly ~1 m K2 rads, one K3 in the kitchen 🤔 . My house is ~80 m². The compressor often shuts off briefly and then cycles on again. When the flow temp is set <35 °C it will do this all day. It can run continuously at >40 °C flow temp but I only do this when electricity is really cheap 😆
> 
> Anyway I think I am diluting this thread so I will butt out 🙂

no worries, info is probably helpful for others in the same situation.
It does sounds like your heatpump is oversized. But then you can do the reverse, lower the flow rate (pump speed), then you need to increase/decrease the feed temp to compensate for the lower flow rate (tweak it that it ask for more power). It should stop cycling (because you are going to ask for more output power than the min output power) when you found the sweetspot. You will have a bit more consumption, but thats better than a cycling heatpump

**Comment by gekkekoe** [comment]
what speed is it now?

Increase pump wil add a couple of watts of consumption, but you can lower feed temp which will reduce a lot more watts of consumption.

**Reply by razorclam** [reply]
Good question, I will check

**Reply by gekkekoe** [reply]
check flow rate afterwards, if it did not increase set it back.

**Comment by jamessimpson007** [comment]
I faffed around with different filters in HA for ages but couldn't get it to work nicely without a significant amount of skew/error especially when the defrost spikes are frequent. In the end I had much better results using `binary_sensor.ecodan_heatpump_defrost` to help, as mentioned above. I combined this with a second templated binary_sensor to introduce a delay to cover the temperature spikes. On my system these run approx 3 min behind the `on` state of `binary_sensor.ecodan_heatpump_defrost` if the circulation pump is running, and up to 25 mins behind if it's not. Here is the config in case it helps anyone else. 

```
- binary_sensor:
    - name: "Defrost Running with Delay"
      unique_id: defrost_running_delay
      icon: mdi:snowflake-melt
      device_class: running
      delay_off: >
        {% if is_state('binary_sensor.ecodan_heatpump_water_pump_status', "off") %}  
          {% set mm = 35 %}
        {% else %} 
          {% set mm = 3 %}
        {% endif %}   
        00:{{mm}}:00
      state: >
        {{ is_state('binary_sensor.ecodan_heatpump_defrost', 'on') }}  

- sensor:
    - name: Ecodan Heat Pump Outside Temperature Filtered 
      unique_id: ecodan_heat_pump_outside_temperature_filtered
      state: >-
        {% set filtered = states('sensor.ecodan_heatpump_outside_temp') | float(6)  %}
        {{ filtered if is_state('binary_sensor.defrost_running_with_delay', "off") else this.state }}
      availability: |
        {{ not states( 'sensor.ecodan_heatpump_outside_temp' ) in ['unavailable', 'unknown']  }}
      state_class: measurement
      device_class: temperature 
      unit_of_measurement: "°C"
```

**Reply by gekkekoe** [reply]
other than fancy graphs, why are you guys wanting to remove the spikes? During defrost AA does not use the outside temp anyway and there's only a theoretical max window of 5m where it sees a 'wrong' outside temp after recovering from defrost, but it's fixed 5m later.

Also keep in mind that it is really warmer at sensor location right after defrost. The system is using that location, so it does not make a lot of sense to filter out values... If you want outside weather, get a meteo reading/feed for your gps location? 

It's a different story if you have the sensor fully exposed in direct sun, then It might report too high values, but that cannot be filtered out anyways.

**Reply by jamessimpson007** [reply]
Ha, admittedly fancy graphs is definitely part of it! However I also have a bunch of automations for forecasting and energy usage which use outside temperature. I've also found that my local forecasting meteo readings are a bit variable....

**Reply by gekkekoe** [reply]
aggregate avg temp over 30m or 60m ? I would not use a single temp point for forecasting :)

**Reply by DamijanD** [reply]
For me - this is my only outside temp sensor that I have in HA, that is why I want to see "accurate" values. But if I had anything else then I wouldn't care for defrost spikes.

**Reply by jamessimpson007** [reply]
Ye, after the above setup I've been grouping all values by each hour and taking the mean, which gives me more meaningful trends for my energy forecasting setup.

**Reply by DamijanD** [reply]
this is how I have it:
<img width="1883" height="404" alt="image" src="https://github.com/user-attachments/assets/7e2e8e24-5ff0-4009-9654-308b42fefe8b" />
